aiohttp/web_ws.py

```diff
@@ -549,36 +558,38 @@ async def receive(self, timeout: Optional[float] = None) -> WSMessage:
 
             try:
                 self._waiting = True
-                try:
-                    if receive_timeout:
-                        # Entering the context manager and creating
-                        # Timeout() object can take almost 50% of the
-                        # run time in this loop so we avoid it if
-                        # there is no read timeout.
-                        async with async_timeout.timeout(receive_timeout):
-                            msg = await self._reader.read()
-                    else:
+                if receive_timeout:
+                    # Entering the context manager and creating
+                    # Timeout() object can take almost 50% of the
+                    # run time in this loop so we avoid it if
+                    # there is no read timeout.
+                    async with async_timeout.timeout(receive_timeout):
                         msg = await self._reader.read()
-                    self._reset_heartbeat()
-                finally:
-                    self._waiting = False
-                    if self._close_wait:
-                        set_result(self._close_wait, None)
+                else:
+                    msg = await self._reader.read()
+                self._reset_heartbeat()
             except asyncio.TimeoutError:
                 raise
             except EofStream:
                 self._close_code = WSCloseCode.OK
-                await self.close()
+                await self._close_from_receive(WSCloseCode.OK)
                 return WS_CLOSED_MESSAGE
             except WebSocketError as exc:
                 self._close_code = exc.code
-                await self.close(code=exc.code)
+                await self._close_from_receive(exc.code)
                 return WSMessageError(data=exc)
             except Exception as exc:
                 self._exception = exc
                 self._set_closing(WSCloseCode.ABNORMAL_CLOSURE)
-                await self.close()
+                await self._close_from_receive(WSCloseCode.OK)
```

Comment:
This is a bit unexpected, but this is how is currently works. Maybe something to reconsider later.